runtime/wasm/src/host_exports.rs

```diff
@@ -142,17 +142,20 @@ impl HostExports {
             entity_type,
             entity_id,
         };
-
-        state.entity_cache.set(key.clone(), Entity::from(data));
-
-        if self.store.uses_relational_schema(&self.subgraph_id)? {
-            // Validate the changes against the subgraph schema
+        let entity = Entity::from(data);
+        state.entity_cache.set(key.clone(), entity.clone());
+
+        // Validate the changes against the subgraph schema.
+        // If the set of fields we have is already valid, avoid hitting the DB.
+        let schema = self.store.input_schema(&self.subgraph_id)?;
+        if validate_entity(&schema.document, &key, &entity).is_err()
```

Comment:
You could save a clone of the entity if you did the first `validate_entity` before the `entity_cache.set`.

Reply:
Ah yes I should do that.

Reply:
Done, merging.